maplibre/src/tile/tile_repository.rs

```diff
@@ -97,40 +114,44 @@ impl TileRepository {
             .map(|results| results.layers.iter())
     }
 
-    /// Removes all the cached tessellate layers that are not contained within the given
-    /// layers hashset.
-    pub fn retain_missing_layer_names(
-        &self,
-        coords: &WorldTileCoords,
-        layers: &mut HashSet<String>,
-    ) {
-        if let Some(cached_tile) = coords.build_quad_key().and_then(|key| self.tree.get(&key)) {
-            let tessellated_set: HashSet<String> = cached_tile
-                .layers
-                .iter()
-                .map(|tessellated_layer| tessellated_layer.layer_name().to_string())
-                .collect();
-
-            layers.retain(|layer| !tessellated_set.contains(layer));
+    /// Create a new tile.
+    pub fn create_tile(&mut self, coords: &WorldTileCoords) -> bool {
+        if let Some(entry) = coords.build_quad_key().map(|key| self.tree.entry(key)) {
+            match entry {
+                btree_map::Entry::Vacant(entry) => {
+                    entry.insert(StoredTile::new());
+                }
+                _ => {}
+            }
         }
+        true
     }
 
-    /// Checks if a layer is missing from the given layers set at the given coords.
-    pub fn is_layers_missing(&self, coords: &WorldTileCoords, layers: &HashSet<String>) -> bool {
-        if let Some(cached_tile) = coords.build_quad_key().and_then(|key| self.tree.get(&key)) {
-            let tessellated_set: HashSet<&str> = cached_tile
-                .layers
-                .iter()
-                .map(|tessellated_layer| tessellated_layer.layer_name())
-                .collect();
+    /// Set the status to success
+    pub fn success(&mut self, coords: &WorldTileCoords) {
+        if let Some(cached_tile) = coords
+            .build_quad_key()
+            .and_then(|key| self.tree.get_mut(&key))
+        {
+            cached_tile.status = TileStatus::Success;
+        }
+    }
 
-            for layer in layers {
-                if !tessellated_set.contains(layer.as_str()) {
-                    return true;
-                }
-            }
+    /// Checks if a layer has been fetched.
+    pub fn fail(&mut self, coords: &WorldTileCoords) {
+        if let Some(cached_tile) = coords
+            .build_quad_key()
+            .and_then(|key| self.tree.get_mut(&key))
+        {
+            cached_tile.status = TileStatus::Failed;
+            cached_tile.retry += 1;
+        }
+    }
 
-            return false;
+    /// Checks if a layer has been fetched.
+    pub fn needs_fetching(&self, coords: &WorldTileCoords) -> bool {
+        if let Some(cached_tile) = coords.build_quad_key().and_then(|key| self.tree.get(&key)) {
+            return cached_tile.status == TileStatus::Failed && cached_tile.retry < 3;
```

Comment:
```suggestion
            return cached_tile.status == TileStatus::Failed && cached_tile.retry < 3;
```

The magic value 3 should be configurable.